### ISSUE TITLE
NVDA UI pop-up dialog focus issues

### ISSUE THREAD

**Issue by michaelDCurran** [nvaccess/test_nvdaTracImport#106]
**Reported by jteh on 12 Feb 2008 08:55 UTC**
**Blocking #7**
There are several issues with regard to foreground focus and NVDA UI pop-up dialogs:
  * While the NVDA UI window is open, pop-up dialogs such as the Exit NVDA dialog intermittently do not come to the foreground so that it is necessary to press alt+tab to access them, as reported in #7.
  * If the NVDA UI window is open, after an NVDA pop-up dialog is dismissed, the focus returns to the NVDA UI window instead of to the window which had the focus before the dialog was popped up.

These are proving to be a little difficult to solve. Some thoughts:
  * All dialogs, including those generated by scriptUI, should have a parent of the NVDA UI window instead of None.
  * When bringing up a scriptUI dialog, if the main UI window is not currently shown, perhaps Show() should be called which should bring it into focus, similar to the code for the Exit NVDA dialog.
  * It seems that winUser.setForegroundWindow() does work quite nicely, despite my earlier findings. This may solve many problems, particularly the second issue described above.
  * All of this fiddly code should be abstracted in some way; i.e. the code to prepare for showing a dialog and the code to restore focus after displaying the dialog.

**Comment by michaelDCurran** [comment]
**Modified by jteh on 14 Feb 2008 21:56 UTC**

**Comment by michaelDCurran** [comment]
**Comment by jteh on 29 Feb 2008 05:45 UTC**
More investigation revealed that !SetForegroundWindow() may not work if another process is in the foreground. This is by design, but it means that we can't use this to reliably bring the NVDA window into focus.

The GUI has needed a complete makeover for a while now, so rather than try to fix this issue, I decided to completely rework the GUI. This was committed in r1745. There is now no NVDA main window (everything is done from the system tray), which means that this problem should not occur.

Closing as fixed. If there are further problems, reopen this ticket.

**Comment by michaelDCurran** [comment]
**Comment by cmssystem on 22 Jul 2013 20:45 UTC**
I'm having issues with pop up dialog for a crm system and wonder if it is related to this fix. I can't find the changeset r1745 anymore, does anyone know if there is a way i can look at the changeset ?

Best regards
crm system

**Comment by michaelDCurran** [comment]
**Comment by cmssystem on 22 Jul 2013 20:47 UTC**
Replying to [crm system](http://www.crmsystemet.dk):

Never mind. Found the issue.

> I'm having issues with pop up dialog for a crm system and wonder if it is related to this fix. I can't find the changeset r1745 anymore, does anyone know if there is a way i can look at the changeset ?
> 
> Best regards
> crm system
